File: viewer/src/lib/schemas.ts
/**
 * TypeScript mirror of src/rlm_logger/schemas.py.
 * Any change here MUST be kept in sync with the Python side.
 * Zod schemas validate case files loaded from URL / drag-drop at runtime.
 */
import { z } from "zod";

export const ModelInfo = z.object({
  provider: z.string(),
  name: z.string(),
  temperature: z.number().default(0.2),
});
export type ModelInfo = z.infer<typeof ModelInfo>;

export const LogFileEntry = z.object({
  path: z.string(),
  bytes: z.number().int(),
  rows: z.number().int(),
  sha256: z.string(),
});

export const TimeWindow = z.object({
  start: z.string(),
  end: z.string(),
});

export const LogsManifest = z.object({
  files: z.array(LogFileEntry),
  time_window: TimeWindow,
  total_rows: z.number().int(),
});
export type LogsManifest = z.infer<typeof LogsManifest>;

export const EvidenceLine = z.object({
  event_id: z.string().nullable().optional(),
  file: z.string(),
  line: z.number().int(),
  ts: z.string(),
  service: z.string(),
  level: z.string(),
  text_redacted: z.string(),
  context_before: z.array(z.string()).max(3).default([]),
  context_after: z.array(z.string()).max(3).default([]),
  why: z.string(),
  is_key: z.boolean().default(false),
});
export type EvidenceLine = z.infer<typeof EvidenceLine>;

export const BlastRadius = z.object({
  duration_minutes: z.number().nullable().optional(),
  window_start: z.string().nullable().optional(),
  window_end: z.string().nullable().optional(),
}).passthrough();

export const IncidentReport = z.object({
  root_cause: z.string(),
  blast_radius: z.union([z.string(), BlastRadius]),
  evidence: z.array(EvidenceLine),
  remediation: z.string(),
  confidence: z.number().min(0).max(1),
  confidence_rationale: z.string(),
  unknowns: z.array(z.string()).default([]),
});
export type IncidentReport = z.infer<typeof IncidentReport>;

export const ToolName = z.enum([
  "schema",
  "top_errors",
  "search",
  "around",
  "trace",
  "submit_incident_report",
  "llm_query",
]);
export type ToolName = z.infer<typeof ToolName>;

export const Step = z.object({
  step: z.number().int(),
  tool: ToolName,
  args: z.record(z.unknown()).default({}),
  stdout_excerpt: z.string().default(""),
  stderr_excerpt: z.string().default(""),
  elapsed_ms: z.number().int().default(0),
  report_delta: z.record(z.unknown()).nullable().optional(),
});
export type Step = z.infer<typeof Step>;

export const Usage = z.object({
  llm_calls: z.number().int().default(0),
  tool_calls: z.number().int().default(0),
  wall_clock_s: z.number().default(0),
  input_tokens: z.number().int().default(0),
  output_tokens: z.number().int().default(0),
});

export const GroundTruth = z.object({
  incident_id: z.string().nullable().optional(),
  root_cause: z.string(),
  evidence_event_ids: z.array(z.string()).default([]),
  blast_radius: z.union([z.record(z.unknown()), z.string()]).nullable().optional(),
  remediation: z.union([z.record(z.unknown()), z.string()]).nullable().optional(),
  distractors: z.array(z.record(z.unknown())).default([]),
}).passthrough();
export type GroundTruth = z.infer<typeof GroundTruth>;

export const TerminationReason = z.enum([
  "submitted",
  "max_iterations",
  "max_llm_calls",
  "max_wall_clock",
  "aborted",
  "error",
]);

export const CaseFile = z.object({
  version: z.literal("0.1"),
  question: z.string(),
  model: ModelInfo,
  logs_manifest: LogsManifest,
  trajectory: z.array(Step),
  report: IncidentReport.nullable().optional(),
  termination_reason: TerminationReason,
  usage: Usage.default({}),
  ground_truth: GroundTruth.nullable().optional(),
});
export type CaseFile = z.infer<typeof CaseFile>;

File: src/rlm_logger/schemas.py
"""Pydantic v2 schemas for the case-file protocol.

These types define the portable, shareable artifact a run produces.
The Astro viewer at rlm.sh binds directly to this shape, so any change
here is a breaking change to the protocol.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class ModelInfo(BaseModel):
    model_config = ConfigDict(extra="forbid")
    provider: str
    name: str
    temperature: float = 0.2


class LogFileEntry(BaseModel):
    model_config = ConfigDict(extra="forbid")
    path: str
    bytes: int
    rows: int
    sha256: str


class TimeWindow(BaseModel):
    model_config = ConfigDict(extra="forbid")
    start: datetime
    end: datetime


class LogsManifest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    files: list[LogFileEntry]
    time_window: TimeWindow
    total_rows: int


class EvidenceLine(BaseModel):
    """A single log line cited in the report. Carries ±3 lines of context
    so the viewer can render surrounding events without re-reading source."""
    model_config = ConfigDict(extra="forbid")
    event_id: str | None = None
    file: str
    line: int
    ts: datetime
    service: str
    level: str
    text_redacted: str
    context_before: list[str] = Field(default_factory=list, max_length=3)
    context_after: list[str] = Field(default_factory=list, max_length=3)
    why: str
    is_key: bool = False


class BlastRadius(BaseModel):
    model_config = ConfigDict(extra="allow")  # schema varies per incident
    duration_minutes: int | None = None
    window_start: datetime | None = None
    window_end: datetime | None = None


class IncidentReport(BaseModel):
    model_config = ConfigDict(extra="forbid")
    root_cause: str
    blast_radius: str | BlastRadius
    evidence: list[EvidenceLine]
    remediation: str
    confidence: float = Field(ge=0.0, le=1.0)
    confidence_rationale: str
    unknowns: list[str] = Field(default_factory=list)


class Step(BaseModel):
    """One iteration of the agent loop: a tool call and its outcome."""
    model_config = ConfigDict(extra="forbid")
    step: int
    tool: Literal["schema", "top_errors", "search", "around", "trace", "submit_incident_report", "llm_query"]
    args: dict[str, Any] = Field(default_factory=dict)
    stdout_excerpt: str = ""
    stderr_excerpt: str = ""
    elapsed_ms: int = 0
    report_delta: dict[str, Any] | None = None


class Usage(BaseModel):
    model_config = ConfigDict(extra="forbid")
    llm_calls: int = 0
    tool_calls: int = 0
    wall_clock_s: float = 0.0
    input_tokens: int = 0
    output_tokens: int = 0


class GroundTruth(BaseModel):
    """Optional expected answer, for eval grading."""
    model_config = ConfigDict(extra="allow")
    incident_id: str | None = None
    root_cause: str
    evidence_event_ids: list[str] = Field(default_factory=list)
    blast_radius: dict[str, Any] | str | None = None
    remediation: dict[str, Any] | str | None = None
    distractors: list[dict[str, Any]] = Field(default_factory=list)


TerminationReason = Literal["submitted", "max_iterations", "max_llm_calls", "max_wall_clock", "aborted", "error"]


class CaseFile(BaseModel):
    """Top-level artifact. One run = one case file."""
    model_config = ConfigDict(extra="forbid")
    version: Literal["0.1"] = "0.1"
    question: str
    model: ModelInfo
    logs_manifest: LogsManifest
    trajectory: list[Step]
    report: IncidentReport | None = None
    termination_reason: TerminationReason
    usage: Usage = Field(default_factory=Usage)
    ground_truth: GroundTruth | None = None
